discovery_engine_core/core/src/stack/filters.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{document::Document, engine::GenericError};
+use std::collections::HashSet;
+use url::Url;
+use xayn_discovery_engine_providers::Article;
+
+pub(crate) trait ArticleFilter {
+    fn apply(current: &[Document], articles: Vec<Article>) -> Result<Vec<Article>, GenericError>;
+}
+
+struct DuplicateFilter;
+
+impl ArticleFilter for DuplicateFilter {
+    fn apply(
+        current: &[Document],
+        mut articles: Vec<Article>,
+    ) -> Result<Vec<Article>, GenericError> {
+        let mut urls = HashSet::new();
+        for doc in current {
+            urls.insert(doc.resource.url.as_str());
+        }
```

Comment:
the hashset could be collected directly from an iterator to avoid potential reallocations when it grows in the loop:
```suggestion
        let urls = current
            .iter()
            .map(|doc| doc.resource.url.as_str())
            .collect::<HashSet<_>>();
```